### DISCUSSION TITLE
Minio broken since upgrade to 0.67

### DISCUSSION THREAD

**Discussion by sarg3nt** [outline/outline#4690] [Self Hosting] [answered]
I had everything working in 0.66, including uploading files to minio but after the upgrade to 0.67 file uploads are now failing.
I'm having a dickins of a time turning on any kind of debug to determine where  the problem is.  If I set DEBUG=* then there's a ton of data but I can't find an error having to do with S3.  With DEBUG disabled there are no errors output at all.  if I set NODE_ENV=debug then the app doesn't start correctly and throws these errors
```sh
$ node ./build/server/index.js
info: [lifecycle] Starting collaboration service
info: [lifecycle] Starting websockets service
info: [lifecycle] Starting worker service
info: [lifecycle] Starting web service
info: [lifecycle] Listening on http://localhost:443
info: [multiplayer] New connection to "document.0a6e6d2f-37c4-4983-8843-b1f4c5eeeb88"
info: [database] Document 0a6e6d2f-37c4-4983-8843-b1f4c5eeeb88 is in database state
info: [multiplayer] Loaded document "document.0a6e6d2f-37c4-4983-8843-b1f4c5eeeb88" {"userId":"970cb7eb-86f1-4b40-a9e2-c719f437a978"}
info: [multiplayer] Closed connection to "document.0a6e6d2f-37c4-4983-8843-b1f4c5eeeb88" {"userId":"970cb7eb-86f1-4b40-a9e2-c719f437a978"}
info: [multiplayer] New connection to "document.0a6e6d2f-37c4-4983-8843-b1f4c5eeeb88"
node:fs:585
  handleErrorFromBinding(ctx);
  ^

Error: ENOENT: no such file or directory, open '/opt/outline/build/server/static/error.dev.html'
    at Object.openSync (node:fs:585:3)
    at Object.readFileSync (node:fs:453:35)
    at readErrorFile (/opt/outline/build/server/onerror.js:39:58)
    at Object.html (/opt/outline/build/server/onerror.js:203:16)
    at Object.app.context.onerror (/opt/outline/build/server/onerror.js:132:12)
    at onerror (/opt/outline/node_modules/koa/lib/application.js:165:32) {
  errno: -2,
  syscall: 'open',
  code: 'ENOENT',
  path: '/opt/outline/build/server/static/error.dev.html'
}
info: [lifecycle] Starting collaboration service
info: [lifecycle] Starting websockets service
info: [lifecycle] Starting worker service
info: [lifecycle] Starting web service
info: [lifecycle] Listening on http://localhost:443
node:fs:585
  handleErrorFromBinding(ctx);
  ^

Error: ENOENT: no such file or directory, open '/opt/outline/build/server/static/error.dev.html'
    at Object.openSync (node:fs:585:3)
    at Object.readFileSync (node:fs:453:35)
    at readErrorFile (/opt/outline/build/server/onerror.js:39:58)
    at Object.html (/opt/outline/build/server/onerror.js:203:16)
    at Object.app.context.onerror (/opt/outline/build/server/onerror.js:132:12)
    at onerror (/opt/outline/node_modules/koa/lib/application.js:165:32) {
  errno: -2,
  syscall: 'open',
  code: 'ENOENT',
  path: '/opt/outline/build/server/static/error.dev.html'
}
```
And just continues to loop like that.

Everything is running as containers and Minio is working.
How can I get debug output that is useful for determining what is going on? OR is there a known issue with 0.67.1 and Minio?

**Comment by tommoor** [comment, marked as answer]
It appears as though you're running development mode? Ensure `NODE_ENV=production` is set. Are you running the official docker image?

**Reply by sarg3nt** [reply]
I was in production mode.  I switched it to development mode to try and get some output to help with why minio was working.  It's back in production now.  
I version locked on :0.67.1 since the auto upgrade 0.67 broke me with the redis user issue.  

I'm trying to get email working too and having issues there as well.  Seeing this in the `DEBUG:email` logs.  Not actually helpful unfortunately.  
```
{"error":"22764511017776:error:1408F10B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:331:\n","stack":"Error: 22764511017776:error:1408F10B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:331:\n","level":"error","message":"Error sending email to dave@sarg3.net"}
{"error":"22764511017776:error:1408F10B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:331:\n","stack":"Error: 22764511017776:error:1408F10B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:331:\n","level":"error","message":"Error processing task in EmailTask"}
{"templateName":"DocumentNotificationEmail","notificationId":"d52f0458-8933-4a7b-8dac-03d3c5941a6a","props":{"to":"dave@sarg3.net","eventName":"published","documentId":"c76d0de7-1573-4b63-b0ab-d9ca7326e0fc","teamUrl":"https://outline.sarg3.net","actorName":"authentik Default Admin","collectionName":"Homelab","unsubscribeUrl":"https://outline.sarg3.net/api/notificationSettings.unsubscribe?token=e72355e7fece2914905ca48caf157113344a761d8126f00b1a1e2977aa0098c2&id=47fc4609-01ee-4762-9ad2-0ff3010af7b2"},"label":"worker","level":"info","message":"EmailTask running"}
{"label":"email","level":"info","message":"Sending email \"“Test for email 3” published\" to dave@sarg3.net"}
{"error":"22764511017776:error:1408F10B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:331:\n","stack":"Error: 22764511017776:error:1408F10B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:331:\n","level":"error","message":"Error sending email to dave@sarg3.net"}
{"error":"22764511017776:error:1408F10B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:331:\n","stack":"Error: 22764511017776:error:1408F10B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:331:\n","level":"error","message":"Error processing task in EmailTask"}
```

**Reply by tommoor** [reply]
So what does the file upload failure look like?

**Reply by sarg3nt** [reply]
When I drag an image into the editor or try to paste a copied image, it disappears almost immediately and a toast pops up that says "i Upload failed"
See attached image.
![image](https://user-images.githubusercontent.com/3372486/212242156-35ea1059-898c-4f44-9067-2669fe20b687.png)

**Reply by tommoor** [reply]
I would like to get to the bottom of it as there were minimal changes in this area in the release. If you can provide the full logs on both the client and server that should be enough.

For the email errors you can search the forum here, it's been solved a couple of times with the right env config.

**Reply by sarg3nt** [reply]
@tommoor  Well, now I'm even more confused.
Your prompt to look at browser logs helped.  I was getting Same Origion policy errors for CORS.  I remembered changing some security settings in Traefik (the reverse proxy I'm using in front of Outline) so I investigated that.  I had added the following:
```yaml
        frameDeny: true
        addVaryHeader: true
```
So I commented those out, restarted Traefik and Ctrl+F5 refreshed Firefox and **boom** it started working!
At work we have a "turn it on, turn it off and back on again" policy when troubleshooting, so I uncommented those two values, restarted traefik and hard refreshed the browser and it  . . . .kept working, no CORS errors . . .what?

As for the email issue, I have searched the forum here and have not turned up anything of use.
I commented on [this ](https://github.com/outline/outline/discussions/1901)discussion as it seems similar but it doesn't look like @routerino has had any luck getting it answered.
I have apple mail configured with several other hosted apps with success so I'm fairly confident my settings are solid.  The StartTLS vs SSL is usually the sticking point.

**Reply by sarg3nt** [reply]
It stopped working again so I commented out those two values, restarted traefik, outline and minio and it started working again.
I'm going to leave it as is for a while and test it every now and then and see if it stays stable.

In other news, the thing that got me started down this road was that I was trying to change the logo in the top left logo.  I can upload it now but the logo does not display.  
Getting this error:
`GET https://minio.sarg3.net/outline-data/public/970cb7eb-86f1-4b40-a9e2-c719f437a978/88bddf65-54d6-4bc7-8542-cf37823116e4/Sarge.png
[HTTP/2 403 Forbidden 4ms]`

Images in documents work fine, including uploading new ones.  I've switched docs and gone back and they do load.  I've force refreshed the browser and they still load.  Not sure why the logo is "special"

Here's the minio policy I'm using.  Got it from a walkthrough on using minio with Outline

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "s3:GetBucketLocation",
                "s3:ListBucket",
                "s3:ListBucketMultipartUploads"
            ],
            "Resource": [
                "arn:aws:s3:::outline-data"
            ]
        },
        {
            "Effect": "Allow",
            "Action": [
                "s3:AbortMultipartUpload",
                "s3:DeleteObject",
                "s3:GetObject",
                "s3:ListMultipartUploadParts",
                "s3:PutObject"
            ],
            "Resource": [
                "arn:aws:s3:::outline-data/*"
            ]
        }
    ]
}
```

**Reply by tommoor** [reply]
The logo and user avatars are special in that they have public permissions to allow them to display on unauthenticated pages. It could be that the bucket has a setting that disables this

**Reply by sarg3nt** [reply]
@tommoor , found it.  The [docs](https://wiki.generaloutline.com/s/hosting/doc/aws-s3-N4M0T6Ypu7 ) say to make the bucket public, which is about the worst thing you can do for security with s3 / minio so I ignored it.  Now that I know what needed public access I went to the minio interface, the bucket, clicked on "Anonymous" and added a read only rule to the `public/` path and that fixed it.
Would probably be good to add docs for minio and the proper way to make that path public.  The current wording is really really bad practice and doesn't make your organization look great from a security view.

![image](https://user-images.githubusercontent.com/3372486/212445557-3cf261c1-d754-4fe6-a702-1679bbd10825.png)

**Reply by tommoor** [reply]
The docs say to turn "Block public access" off, not to make the bucket public. This is a specific setting for S3 though. 

It would be nice to have some docs for Minio but I don't know much about managing that software so it's been left as an exercise for the end user up until now.

**Reply by sarg3nt** [reply]
Ahh, my apologies, I read that wrong.  Minio doesn't have that option but does have an easy way to make the entire bucket public so I thought that's what it was saying.  Thank you for the clarification.
I could write up some docs for Minio if you'd like.  I was planning on doing so for my own install notes anyway. I'm not sure when I will get to them, could be a few days.

**Reply by tommoor** [reply]
No worries, if you'd like to start something that would be great – I'll happily transfer that over to the official docs as it's ready.

**Reply by tclemans** [reply]
@sarg3nt are your MinIO install notes available?

**Reply by MDXDave** [reply]
Change the access policy of your bucket to custom and make only the public folder available:

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "PublicFolder",
            "Effect": "Allow",
            "Principal": {
                "AWS": [
                    "*"
                ]
            },
            "Action": [
                "s3:GetObject"
            ],
            "Resource": [
                "arn:aws:s3:::<<BUCKETNAME>>/public/*"
            ]
        }
    ]
}
```
